Browse View Enhancements

Ben2356 opened the pull request:
With the [enqueued view enhancements rework](https://github.com/hufman/AndroidAutoIdrive/pull/128) came the system for displaying cover art for lazily loading cover art into a list view which I have extended to the "Browse" hmiState to enhance the presentation. The following features/fixes have been added: 
- Added the cover art for each of the browse items which are lazily loaded into the browse pages. If there is no cover art for the browse item then the song/folder icon will be displayed in its place
- Added a line to the browse items that displays additional information about the browse item if there is one present
- The back functionality will now be able to be used for when there are greater than 2 levels of pages (prior behavior would go back to the PlaybackView when the user tried to go back from a page that was greater than 2 levels, for example if the user went to "Your Library" -> "Playlists" and then back out of the "Playlists" browse page)
- Android app has been updated to show the cover art images where applicable otherwise falling back on the song/folder icon

BrowseView in action  (gif compression and frame rate are terrible to make it into a size that is attachable on github):
![newBrowseView](https://user-images.githubusercontent.com/11298573/97794860-4b452380-1bcd-11eb-99ed-e211810dec22.gif)

PhoneUI in action:
![browseViewPhoneUI](https://user-images.githubusercontent.com/11298573/97794784-29976c80-1bcc-11eb-8542-38a29ce34e14.jpg)

Also this PR should address this issue: https://github.com/hufman/AndroidAutoIdrive/issues/109

hufman commented on app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowsePageView.kt:
```diff
@@ -88,35 +98,14 @@ class BrowsePageView(val state: RHMIState, val musicImageIDs: MusicImageIDs, val
 	fun initWidgets(inputState: RHMIState) {
 		folderNameLabel.setVisible(true)
+
 		// handle action clicks
```
Great refactoring, thanks!

hufman commented on app/src/test/java/me/hufman/androidautoidrive/TestMusicApp.kt:
```diff
@@ -1265,28 +1424,67 @@ class TestMusicApp {
 				MusicMetadata("bonusFolder1", title = "BonusFolder1", browseable = true, playable = false),
 				browseView.pageStack[1].previouslySelected)
 
+		// select a deeper folder (bonusFolder2:BonusFolder2, to show in BROWSE2)
```
Good, make sure the wraparound test is still present

Ben2356 replied:
Yea I have it so now it's browsing a level deeper which means it will hit that wraparound case BROWSE1 -> BROWSE2 -> BROWSE3 -> BROWSE2, I spent way too much time on getting that test to work all because I messed up a single focus event 😆

hufman commented on app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowseView.kt:
```diff
@@ -32,17 +34,20 @@ class BrowseView(val states: List<RHMIState>, val musicController: MusicControll
 
 
-	fun initWidgets(playbackView: PlaybackView, inputState: RHMIState) {
+	fun initWidgets(playbackView: PlaybackView, inputState: RHMIState, musicApp: MusicApp) {
```
The `initWidgets` function is used to link up a bunch of states together after they have all be constructed. The MusicApp could instead be passed to the BrowseView's constructor

Ben2356 replied:
Ah that would be a much better spot for it, reverted the changes to the initWidgets() method and added the param to the BrowseView constructor

hufman commented on app/src/main/java/me/hufman/androidautoidrive/carapp/music/MusicApp.kt:
```diff
@@ -36,6 +36,7 @@ class MusicApp(val securityAccess: SecurityAccess, val carAppAssets: CarAppResou
 	var appListViewVisible = false
 	var playbackViewVisible = false
 	var enqueuedViewVisible = false
+	var browseViewVisible = false
```
This is fine for following the pattern, but I should change this (to match a pattern I've developed in Notifications) where each view has its own visible flag, instead of tracking these flags in this parent object. No need to include that change in this PR though, just musings :)

Ben2356 replied:
I do like that new structure! It sounds like eventually going to need a PR to sync all the states to use a similar system :)

hufman commented on app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowsePageView.kt:
```diff
@@ -166,13 +159,33 @@ class BrowsePageView(val state: RHMIState, val musicImageIDs: MusicImageIDs, val
+						val checkmarkIcon = if (previouslySelected == item) checkmarkIcon else ""
+						val coverArt = item.coverArt
+						val coverArtImage =
+								if (coverArt != null)
```
Can these if/else clauses have brackets, to match the rest of the code?

Ben2356 replied:
Yeah sorry I must have missed that, updated to have brackets

hufman commented on app/src/main/java/me/hufman/androidautoidrive/carapp/music/MusicApp.kt:
```diff
@@ -339,6 +343,10 @@ class MusicApp(val securityAccess: SecurityAccess, val carAppAssets: CarAppResou
 			enqueuedView.redraw()
 		}
+		if (browseViewVisible) {
+			browseView.currentPage?.redraw()
```
I think this should be changed to call a `browseView.redraw()` function, allowing the browseView's currentPage to be private and reducing coupling slightly.

Ben2356 replied:
I agree, added a redraw method to BrowseView and calling that here

hufman commented:
Would you be able to rewrite the commit history to have a single commit per change? I think there would just be the two if I'm reading it right:

- Fixes nested browsing pages
- Implements browse coverart for #109

Thank you for this excellent work! I'm happy that the SpotifyMusicMetadata object worked out so easily!

hufman commented on app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowseView.kt:
```diff
@@ -87,16 +92,24 @@ class BrowseView(val states: List<RHMIState>, val musicController: MusicControll
 			pushBrowsePage(null, stateId)
 		}
 		// show the content for the page that we are showing
-		val currentPage = pageStack.last()
```
Perhaps keep this `val currentPage` in place for the nonnull guarantee, and then add a `this.currentPage = currentPage` after.

Ben2356 replied:
Yeah that would be cleaner, reverted these changes and added that member var assignment to the currentPage

Ben2356 commented:
> 
> 
> Would you be able to rewrite the commit history to have a single commit per change? I think there would just be the two if I'm reading it right:
> 
>     * Fixes nested browsing pages
> 
>     * Implements browse coverart for #109
> 
> 
> Thank you for this excellent work! I'm happy that the SpotifyMusicMetadata object worked out so easily!

Rebased and squashed down to 2 commits, usually I wait to do the squashing until the end that way I have commit history in case I need to revert or something haha.

You're very welcome!! I'm glad too, it was a lot easier than I thought it would be which was great
